esp/xslt/esdl2cpp_srvbasehpp.xslt

```diff
@@ -0,0 +1,273 @@
+<!--
+##############################################################################
+#    HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+##############################################################################
+-->
+
+<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/" xmlns:http="http://schemas.xmlsoap.org/wsdl/http/" xmlns:mime="http://schemas.xmlsoap.org/wsdl/mime/" xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/">
+    <xsl:output method="text" omit-xml-declaration="yes" indent="no"/>
+
+    <xsl:template match="esxdl">
+<xsl:variable name="servicename"><xsl:value-of select="EsdlService/@name"/></xsl:variable>
+<xsl:variable name="upperService"><xsl:value-of select="translate(EsdlService/@name, 'abcdefghijklmnopqrstuvwxyz', 'ABCDEFGHIJKLMNOPQRSTUVWXYZ')"/></xsl:variable>
+<xsl:text>#ifndef </xsl:text><xsl:value-of select="$upperService"/>SERVICEBASE_HPP__
+<xsl:text>#define </xsl:text><xsl:value-of select="$upperService"/>SERVICEBASE_HPP__
+
+#include "jlib.hpp"
+#include "jptree.hpp"
+#include "jarray.hpp"
+#include "primitivetypes.hpp"
+
+using namespace std;
+using namespace cppplugin;
+
+class EsdlContext : public CInterface, implements IInterface
+{
+public:
+    IMPLEMENT_IINTERFACE;
+
+    Owned&lt;PString&gt; username;
+    Owned&lt;Integer&gt; clientMajorVersion;
+    Owned&lt;Integer&gt; clientMinorVersion;
+};
+
+<xsl:apply-templates select="EsdlEnumType"/>
+<xsl:apply-templates select="EsdlStruct"/>
+<xsl:apply-templates select="EsdlRequest"/>
+<xsl:apply-templates select="EsdlResponse"/>
+<xsl:apply-templates select="EsdlService"/>
+// User need to implement this function
+extern "C" <xsl:value-of select="$servicename"/>ServiceBase* create<xsl:value-of select="$servicename"/>ServiceObj();
+
+<xsl:text>
+#endif</xsl:text>
+    </xsl:template>
+
+    <xsl:template match="EsdlStruct|EsdlRequest|EsdlResponse">
+class <xsl:value-of select="@name"/><xsl:choose><xsl:when test="@base_type"> : public <xsl:value-of select="@base_type"/></xsl:when><xsl:otherwise> : public CInterface, implements IInterface</xsl:otherwise></xsl:choose>
+{
+public:
+    IMPLEMENT_IINTERFACE;
+
+<xsl:apply-templates select="EsdlElement|EsdlArray|EsdlEnum"/><xsl:text>};
+</xsl:text>
+    </xsl:template>
+
+    <xsl:template name="outputCppPrimitive">
```

Comment:
I tried it and found it can't be easily done. The transform seems to require the full path of the included file, relative path doesn't work. But the full path isn't known until after installation. We could do something fancy like generating the XSLT file at installation time or run time, but I don't think it's worth the effort.